### PR TITLE
Add ansible.builtin.pipe action plugin

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY

It allows piping the output a local command to a remote command.

The reason for this was transferring a docker image built on the local machine to a remote machine using `docker save image:tag | docker load`. To my knowledge, this was previously only possible by saving the image locally, copying it over to the remote host, loading the image on the remote host, and cleaning up the `.tar` images left on both sides.

Now, all those steps can be replaced by 

```yaml
ansible.builtin.pipe:
        local: "docker save image:tag"
        remote: docker load
```

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

Plugin: `ansible.builtin.pipe`

##### ADDITIONAL INFORMATION

The module exists only because I'm not sure how else to document the plugin.

### CODE REVIEW

File: lib/ansible/modules/pipe.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+- ansible.modules.script
+author:
+- LoveIsGrief
+'''
+
+EXAMPLES = r'''
+- name: Copy file with owner and permissions
+  ansible.builtin.pipe:
+    src: docker save {{ image }} 
+    dest: docker load
```
Comment (flowerysong) [inline lib/ansible/modules/pipe.py]: This example does not match the documented options.

File: lib/ansible/modules/pipe.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+'''
+
+RETURN = r'''
+'''
```
Comment (flowerysong) [inline lib/ansible/modules/pipe.py]: Return values should be documented.

File: lib/ansible/modules/pipe.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        ),
+        supports_check_mode=False,
+    )
+    module.fail_json(
+        msg='Cannot be executed directly. All work is done in the pipe action',
+    )
+
+
+if __name__ == '__main__':
+    main()
```
Comment (flowerysong) [inline lib/ansible/modules/pipe.py]:
```suggestion
```

Documentation for action plugins should not include unnecessary module implementations. See e.g. https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/reboot.py

File: lib/ansible/plugins/action/pipe.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        except Exception as e:
+            raise AnsibleError('Could not execute remote command: %s' % to_native(e))
+
+        local_proc.wait(timeout=5)
+        return {
+            'src': {
+                'return_code': local_proc.returncode,
+            },
+            'dest': {
+                'return_code': dest_return_code,
```
Comment (flowerysong) [inline lib/ansible/plugins/action/pipe.py]:
```suggestion
            'local': {
                'rc': local_proc.returncode,
            },
            'remote': {
                'rc': dest_return_code,
```

These should probably be abbreviated to `rc` for consistency with other modules that return process results, like `command`.

It feels like the `stdout` and `stderr` of the process on the target machine should be returned (and maybe `stderr` from the controller.)

File: lib/ansible/plugins/connection/ssh.py [hunk 5]
```diff
@@ -5,8 +5,11 @@
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 from __future__ import (absolute_import, division, print_function)
+
```
Comment (flowerysong) [inline lib/ansible/plugins/connection/ssh.py]:
```suggestion
```

Unnecessary whitespace change; this boilerplate is not supposed to have a newline between those two lines.

File: lib/ansible/plugins/connection/ssh.py [hunk 6]
```diff
@@ -5,8 +5,11 @@
...
 __metaclass__ = type
 
+import io
+
```
Comment (flowerysong) [inline lib/ansible/plugins/connection/ssh.py]: Imports go below DOCUMENTATION.

File: lib/ansible/plugins/connection/ssh.py [hunk 7]
```diff
@@ -749,9 +752,10 @@ def _send_initial_data(self, fh, in_data, ssh_process):
         it. (The handle must be closed; otherwise, for example, "sftp -b -" will
         just hang forever waiting for more commands.)
         '''
+        if isinstance(in_data, io.IOBase):
+            return self._send_initial_file_data(fh, in_data, ssh_process)
 
         display.debug(u'Sending initial data')
-
```
Comment (flowerysong) [inline lib/ansible/plugins/connection/ssh.py:754]: Unnecessary whitespace change.

**Review by flowerysong — COMMENTED** [review]
Just some general comments; you may want to wait for feedback from the core team on whether they're interested in this feature before spending time addressing them.

**Comment by LoveIsGrief** [comment]
I think they were good comments @flowerysong :+1: . Thanks.

**Comment by bcoca** [comment]
What is the difference vs doing?:
```
- shell: docker load
  args:
    stdin: "{{lookup('pipe', 'docker save image:tag'}}"
```

File: lib/ansible/plugins/connection/ssh.py [hunk 8]
```diff
@@ -768,6 +771,44 @@ def _send_initial_data(self, fh, in_data, ssh_process):
 
         display.debug(u'Sent initial data (%d bytes)' % len(in_data))
 
+    def _send_initial_file_data(self, fh, in_fh, ssh_process):
```
Comment (bcoca) [inline lib/ansible/plugins/connection/ssh.py:773]: This is not needed, the 'in_data' is written to the pipe when provided. This just creates another way of doing the same thing. Also this makes it ssh specific, it should be cross connection plugin (for those that support inline data aka pipes).

Comment (ghost) [reply lib/ansible/plugins/connection/ssh.py:773]: Without this the line right above this `display.debug(u'Sent initial data (%d bytes)' % len(in_data))` won't work because `IOBase` doesn't implement `__len__`.

Additionally, [this line](https://github.com/ansible/ansible/pull/74119/files/2a6817d8a3891b4ce309b56d11dc6bdbcc6310ec#diff-f69a630433cbce2e546a7023edeca91878d27069c0f725c3854d40f7e66b81e0R45)

https://github.com/LoveIsGrief/ansible/blob/ansible-builtin-pipe/lib/ansible/plugins/action/pipe.py#L45

would have to be changed to `in_data=local_proc.stdout.read(),`. I'm quite certain that means reading the entire `stdout` into memory. [`send_initial_data` also calls `to_bytes`](https://github.com/ansible/ansible/pull/74119/files/2a6817d8a3891b4ce309b56d11dc6bdbcc6310ec#diff-38cec806ea1a1ee7c3a286c7865334ecfba7ccc49e21d4e8fb8ec1b17938fda6R759) which will ultimately convert the `IOBase` object to a string.

Edit: Unless it's mandatory, I would like to concentrate on `ssh` first, as I don't have the time yet to test every connector. Would a note in the documentation mentioning the `ssh` restriction be alright? Other connectors could then come later.

**Comment by LoveIsGrief** [comment]
> What is the difference vs doing?:
> 
> ```
> - shell: docker load
>   args:
>     stdin: "{{lookup('pipe', 'docker save image:tag'}}"
> ```

This is what I'm getting

```
fatal: [default]: FAILED! => {
    "msg": "An unhandled exception occurred while running the lookup plugin 'pipe'. Error was a <class 'UnicodeDecodeError'>, original message: 'utf-8' codec can't decode byte 0xc1 in position 7192: invalid start byte"
}
```

due to https://github.com/LoveIsGrief/ansible/blob/ansible-builtin-pipe/lib/ansible/plugins/lookup/pipe.py#L73

That approach still runs into the issue that the entire `stdout` is first loaded into host memory and then processed instead of being streamed as data comes in. Here's a simple example of trying to save the `node:lts` image (~950MB)

```bash
$ ps --pid 7849 -o rss,ucmd
  RSS CMD
1953668 python
````

`1953668KB / 1024 = ~1907MB` in memory. If that's somehow copied somewhere else down the line, the memory could balloon pretty quickly.

**Comment by s-hertel** [comment]
Question About This One
=======================

Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program as this is too niche of a case to include in ansible-core.

You can stop by the development list or IRC, and we'd be glad to talk about it - and we might even be persuaded otherwise!

   * https://groups.google.com/forum/#!forum/ansible-devel
   * #ansible-devel on Freenode IRC
